src/scripts/fetchStatsFromGA.js

```diff
@@ -0,0 +1,418 @@
+        0} ` +
+        `out of total ${
+          report.data.rowCount
+        } rows for ${sourceType} ${docType}` +
+        ` with next pageToken ${report.nextPageToken}`
+    );
+    results[docType] = report.data.rows;
+    nextPageTokens[docType] = report.nextPageToken || -1;
+    hasMore = hasMore || report.nextPageToken != null;
+  });
+
+  return { results, pageTokens: nextPageTokens, hasMore };
+};
+
+const upsertDocStats = async function(body) {
+  const res = await client.bulk({
+    body,
+    refresh: 'true',
+  });
+  if (res.body.errors) {
+    console.error(
+      `bulk upsert failed : ${JSON.stringify(res.body, null, '  ')}`
+    );
+  }
+};
+
+async function fetchReplyUsers(rows) {
+  const replyIds = rows.map(row => parseIdFromRow(row));
+  const {
+    body: {
+      hits: { hits: replies },
+    },
+  } = await client.search({
+    index: 'replies',
+    size: pageSize,
+    body: {
+      query: {
+        ids: { values: replyIds },
+      },
+      _source: {
+        includes: ['userId'],
+      },
+    },
+  });
+  let replyUsers = {};
+  replies.forEach(reply => (replyUsers[reply._id] = reply._source.userId));
+  return replyUsers;
+}
+
+/**
+ * Given stats report for a sourceType and docType, store them in db
+ *
+ * @param {string} sourceType    One of the values defined by statsSources.
+ * @param {string} docType       One of the values defined by docTypes.
+ * @param {array}  rows          Rows of data to process; each row is an object
+    of the form {dimensions: [{field: value}], metrics:[{values:[values]}]}}
+ * @param {Date}   fetchedAt     Timestamp of the fetched time
+ * @param {object} lastParams    Stats for the last processed row from previous
+    call of the same sourceType and docType; It's an object of the form {
+      sourceType: string, docType: string, path: string, date: string,
+      docId: string, visits: number, users: number }
+
+ * @return {object} Return the stats of last processed row, same format as lastParams.
+ */
+const processReport = async function(
+  sourceType,
+  docType,
+  rows,
+  fetchedAt,
+  lastParams
+) {
+  let replyUsers;
+  if (docType === docTypes.REPLY) {
+    replyUsers = await fetchReplyUsers(rows);
+  }
+
+  let bulkUpdates = [];
+  const sourceName = sourceType.toLowerCase();
+
+  rows.forEach(row => {
+    const docId = parseIdFromRow(row);
+    const date = formatDate(row.dimensions[1]);
+    const [visits, users] = row.metrics[0].values.map(v => parseInt(v, 10));
+    const isSameEntry =
+      lastParams && lastParams.date === date && lastParams.docId === docId;
+    let docUserId, stats;
+
+    if (docType === docTypes.REPLY && replyUsers[docId]) {
+      docUserId = replyUsers[docId];
+    }
+
+    stats = {
+      [`${sourceName}Visit`]: parseInt(visits, 10),
+      [`${sourceName}User`]: parseInt(users, 10),
+    };
+
+    if (isSameEntry) {
+      stats = lastParams.stats;
```

Comment:
It would be neat if we have comment noting that this block may modify entries that is pushed into `bulkUpdates` in previous loop.